numpy/core/src/multiarray/item_selection.c

```diff
@@ -1749,6 +1750,275 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
     return NULL;
 }
 
+/*NUMPY_API
+ *Keysort an array providing indices that will sort a collection of arrays
+ *lexicographically.  The first key is sorted on first, followed by the second key
+ *-- requires that arg"merge"sort is available for each sort_key
+ *
+ *Returns None on success, returns NULL on failure
+ */
+NPY_NO_EXPORT PyObject *
+PyArray_KeySort(PyArrayObject *self, PyObject *sort_keys, int axis)
+{
+    PyArrayObject **mps;
+    PyArrayIterObject **its;
+    PyArrayIterObject *sit = NULL;
+    npy_intp n, N, size, i, j;
+    npy_intp astride, *iptr, self_stride;
+    int nd;
+    int needcopy = 0;
+    int elsize;
+    int self_elsize;
+    int maxelsize;
+    int object = 0;
+    PyArray_ArgSortFunc *argsort;
+    NPY_BEGIN_THREADS_DEF;
+
+    if (!PySequence_Check(sort_keys)
+           || ((n = PySequence_Size(sort_keys)) <= 0)) {
+        PyErr_SetString(PyExc_TypeError,
+                "need sequence of keys with len > 0 in keysort");
+        return NULL;
+    }
+    mps = (PyArrayObject **) PyArray_malloc(n * sizeof(PyArrayObject *));
+    if (mps == NULL) {
+        return PyErr_NoMemory();
+    }
+    its = (PyArrayIterObject **) PyArray_malloc(n * sizeof(PyArrayIterObject *));
+    if (its == NULL) {
+        PyArray_free(mps);
+        return PyErr_NoMemory();
+    }
+    for (i = 0; i < n; i++) {
+        mps[i] = NULL;
+        its[i] = NULL;
+    }
+    for (i = 0; i < n; i++) {
+        PyObject *obj;
+        obj = PySequence_GetItem(sort_keys, i);
+        if (obj == NULL) {
+            goto fail;
+        }
+        mps[i] = (PyArrayObject *)PyArray_FROM_O(obj);
+        Py_DECREF(obj);
+        if (mps[i] == NULL) {
+            goto fail;
+        }
+        if (i > 0) {
+            if ((PyArray_NDIM(mps[i]) != PyArray_NDIM(mps[0]))
+                    || (!PyArray_CompareLists(PyArray_DIMS(mps[i]),
+                                       PyArray_DIMS(mps[0]),
+                                       PyArray_NDIM(mps[0])))) {
+                PyErr_SetString(PyExc_ValueError,
+                                "all keys need to be the same shape");
+                goto fail;
+            }
+        }
+        if (!PyArray_DESCR(mps[i])->f->argsort[NPY_STABLESORT]
+                && !PyArray_DESCR(mps[i])->f->compare) {
+            PyErr_Format(PyExc_TypeError,
+                         "item %zd type does not have compare function", i);
+            goto fail;
+        }
+        if (!object
+            && PyDataType_FLAGCHK(PyArray_DESCR(mps[i]), NPY_NEEDS_PYAPI)) {
+            object = 1;
+        }
```

Comment:
Possibly clearer as:
```suggestion
        object = object || PyDataType_FLAGCHK(PyArray_DESCR(mps[i]), NPY_NEEDS_PYAPI));
```
`any_key_is_object` would likely be a better name.